### PR TITLE
Msi 2182 source entity expansion with pickup location attributes

### PR DESCRIPTION
**Pull request by novikor**
### Description (*)
Added the next fields:
- Frontend Name
- Frontend Description (wysiwyg field, label is not shown in BO)
- Implemented admin UI for them;
- Refactored some code;

### Fixed Issues (if relevant)
1. magento-engcom/msi#2182:Source Entity expansion with Pickup Location attributes

### Manual testing scenarios (*)
1. Go to source edit page
2. Make sure pickup location fieldset is not visible until the source is used in the store pickup
3. Check that new fields are present
4. Make sure you are able to save them 

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/InventoryInStorePickup/Test/_files/pickup_location.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
+<?php
```
Comment (swnsma) [inline app/code/Magento/InventoryInStorePickup/Test/_files/pickup_location.php:1]: There is already fixture to make source -> pickup location: `app/code/Magento/InventoryInStorePickup/Test/_files/source_pickup_location_attributes.php`
Do we really need extra one?

File: app/code/Magento/InventoryInStorePickup/Test/Integration/PickupLocation/MapperTest.php [hunk 2]
```diff
@@ -85,86 +88,25 @@ public function testWrongMappingForPickupLocation()
     }
 
     /**
-     * @magentoDataFixture ../../../../app/code/Magento/InventoryApi/Test/_files/source.php
+     * @magentoDataFixture ../../../../app/code/Magento/InventoryInStorePickup/Test/_files/pickup_location.php
      */
     public function testMapPickupLocation()
```
Comment (swnsma) [inline app/code/Magento/InventoryInStorePickup/Test/Integration/PickupLocation/MapperTest.php:93]: What about case: 
`Source.ExtensionAttributes.Field1 -> PickupLocation.ExtensionAttributes.Field2` ?

Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Test/Integration/PickupLocation/MapperTest.php:93]: Why did you not cover it before?)

Okay, I`ll check

**Comment by novikor** [comment]
**TODO:**  parse `FrontendDescription` WYSIWYG content while creating pickup location.
See `Magento\Widget\Model\Template\Filter`

**Comment by novikor** [comment]
@ishakhsuvarov , - the PR is not ready yet.
@swnsma  is going to implement a chain of responsibility for the pickup location mapper,
and I will reuse this logic then

File: app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * @param ExtensionAttributesFactory $extensionAttributesFactory
+     */
+    public function __construct(ExtensionAttributesFactory $extensionAttributesFactory)
+    {
+        $this->extensionAttributesFactory = $extensionAttributesFactory;
+    }
+    /**
+     * @param SourceInterface|DataObject $source
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php]: Let's keep doc block and contract consistent

File: app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+use Magento\Framework\Api\ExtensionAttributesFactory;
+use Magento\Framework\DataObject;
+use Magento\InventoryApi\Api\Data\SourceInterface;
+use Magento\InventoryInStorePickupApi\Api\Data\PickupLocationInterface;
+
+/**
+ * Set store-pickup related source extension attributes
+ */
+class SetExtensionAttributes
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php]: Class name seems to abstract

File: app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        $this->extensionAttributesFactory = $extensionAttributesFactory;
+    }
+    /**
+     * @param SourceInterface|DataObject $source
+     */
+    public function execute(SourceInterface $source): void
+    {
+        $pickupAvailable = $source->getData(PickupLocationInterface::IS_PICKUP_LOCATION_ACTIVE);
+        $frontendName = $source->getData(PickupLocationInterface::FRONTEND_NAME);
+        $frontendDescription = $source->getData(PickupLocationInterface::FRONTEND_DESCRIPTION);
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php]: These fields doesn't seem to be present in Pickup Location entity.

File: app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetPlugin.php [hunk 6]
```diff
@@ -7,49 +7,41 @@
...
+    ) {
+        $this->setExtensionAttributes = $setExtensionAttributes;
     }
 
     /**
      * Enrich the given Source Objects with the In-Store pickup attribute
      *
      * @param SourceRepositoryInterface $subject
-     * @param SourceInterface $source
+     * @param SourceInterface|DataObject $source
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetPlugin.php]: That's cheating. We can't typehint for SourceInterface and rely on Dataobject in reality

File: app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/SaveInStorePickupPlugin.php [hunk 7]
```diff
@@ -7,28 +7,32 @@
 
 namespace Magento\InventoryInStorePickup\Plugin\InventoryApi\SourceRepository;
 
+use Magento\Framework\DataObject;
 use Magento\InventoryApi\Api\Data\SourceInterface;
 use Magento\InventoryApi\Api\SourceRepositoryInterface;
+use Magento\InventoryInStorePickupApi\Api\Data\PickupLocationInterface as Location;
 
 class SaveInStorePickupPlugin
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/SaveInStorePickupPlugin.php:19]: Let's add description while we are here

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
Whole mechanism of working with specific fields of `SourceInteface` via `DataObject` implementation does not seem correct.

Should we research additional ways? Something cleaner like a Resource maybe? Or maybe you have more approaches on your mind?

File: app/code/Magento/InventoryInStorePickup/Test/Integration/PickupLocation/MapperTest.php (hunk 2 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Test/Integration/PickupLocation/MapperTest.php:93]: @swnsma , I afraid it is out of scope. Feel free to do it by yourself.

File: app/code/Magento/InventoryInStorePickup/Test/_files/pickup_location.php (hunk 1 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Test/_files/pickup_location.php:1]: Yes

File: app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php (hunk 5 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php]: Yes, they are not: they are extension attribute codes of `Source`.

I am not sure there is a sense to move these constants elsewhere, - to a separate class with just constants only, for example.

Or it is?

File: app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php (hunk 4 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php]: `InitPickupLocationExtensionAttributes` ?

File: app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetPlugin.php (hunk 6 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetPlugin.php]: Fixed

File: app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/SaveInStorePickupPlugin.php (hunk 7 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/SaveInStorePickupPlugin.php:19]: Done

File: app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php (hunk 3 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Model/Source/SetExtensionAttributes.php]: Done

**Comment by novikor** [comment]
Reminder for myself: display Frontend Description label

File: app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/CreateFromSource.php [hunk 8]
```diff
@@ -47,28 +48,49 @@ public function __construct(
      * @inheritdoc
      * @throws \InvalidArgumentException
      */
-    public function execute(SourceInterface $source, array $map): PickupLocationInterface
+    public function execute(SourceInterface $source, array $map, array $preProcessors): PickupLocationInterface
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/CreateFromSource.php:51]: Why `preProcessors` aren't constructor injection? Adequate type checking and hinting is missing as well.

Edit: Would be resolved in a different PR after additional discussion

File: app/code/Magento/InventoryInStorePickupApi/Model/Mapper/PreProcessorInterface.php [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /**
+     * Process Source Field before pass it to Pickup Location
+     *
+     * @param SourceInterface $source
+     * @param mixed $value Source Field Value
+     *
+     * @return mixed
+     */
+    public function process(SourceInterface $source, $value);
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickupApi/Model/Mapper/PreProcessorInterface.php:27]: Situation where `mixed` type have to be used may be a hint to a non-optimal design. 

In this case, this contract seems incomplete, as Field Name seems to be hardcoded somewhere in the implementation and value has to be mixed.
I also love single responsibility classes, but still, it may be an overkill to introduce a class per field.

File: app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/PreProcessor/FrontendName.php [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+declare(strict_types=1);
+
+namespace Magento\InventoryInStorePickup\Model\PickupLocation\Mapper\PreProcessor;
+
+use Magento\InventoryApi\Api\Data\SourceInterface;
+use Magento\InventoryInStorePickupApi\Model\Mapper\PreProcessorInterface;
+
+class FrontendName implements PreProcessorInterface
+{
+
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/PreProcessor/FrontendName.php:16]: Just an empty line

File: app/code/Magento/InventoryInStorePickup/Model/Source/InitPickupLocationExtensionAttributes.php [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    }
+    /**
+     * @param SourceInterface $source
+     */
+    public function execute(SourceInterface $source): void
+    {
+        if (!$source instanceof DataObject) {
+            return;
+        }
+        $pickupAvailable = $source->getData(PickupLocationInterface::IS_PICKUP_LOCATION_ACTIVE);
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Model/Source/InitPickupLocationExtensionAttributes.php:41]: In the last review I wasn't clear enough.

Just wanted to make sure that this is the best idea we have to load newly introduced fields.

File: app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetListPlugin.php [hunk 12]
```diff
@@ -7,25 +7,24 @@
...
-     * @param ExtensionAttributesFactory $extensionAttributesFactory
+     * @param InitPickupLocationExtensionAttributes $setExtensionAttributes
      */
-    public function __construct(ExtensionAttributesFactory $extensionAttributesFactory)
-    {
-        $this->extensionAttributesFactory = $extensionAttributesFactory;
+    public function __construct(
+        InitPickupLocationExtensionAttributes $setExtensionAttributes
+    ) {
+        $this->setExtensionAttributes = $setExtensionAttributes;
```
Comment (ishakhsuvarov) [inline app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetListPlugin.php:27]: While this seems to be an excessive extraction, it seems to be the bet solution if we don't figure out to load data in a better way.

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

File: app/code/Magento/InventoryInStorePickup/Model/Source/InitPickupLocationExtensionAttributes.php (hunk 11 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Model/Source/InitPickupLocationExtensionAttributes.php:41]: Well, we could do some magic with converting snake_case field name to camelCase, check if the appropriate getter method exists and then call it...

But I don`t like this idea :)

File: app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetListPlugin.php (hunk 12 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Plugin/InventoryApi/SourceRepository/LoadInStorePickupOnGetListPlugin.php:27]: Well, as a variant, it probably would be nice to extract this logic into a chain of responsibility.

File: app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/CreateFromSource.php (hunk 8 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/CreateFromSource.php:51]: Well.
What about creating something kind of `PreProcessorsList` (like `RouterList`), mark it as `@api` and inject into constructor then?

File: app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/PreProcessor/FrontendName.php (hunk 10 above)
Comment (novikor) [reply app/code/Magento/InventoryInStorePickup/Model/PickupLocation/Mapper/PreProcessor/FrontendName.php:16]: That`s really horrible

**Review by ishakhsuvarov — APPROVED** [review]
